pkg/trace/api/listener.go

```diff
@@ -71,32 +90,29 @@ func (e *rateLimitedError) Timeout() bool { return false }
 
 // Accept reimplements the regular Accept but adds rate limiting.
 func (sl *rateLimitedListener) Accept() (net.Conn, error) {
-	if atomic.LoadInt32(&sl.connLease) <= 0 {
-		// we've reached our cap for this lease period, reject the request
+	if atomic.LoadInt32(&sl.lease) <= 0 {
+		// we've reached our cap for this lease period; reject the request
+		atomic.AddUint32(&sl.rejected, 1)
 		return nil, &rateLimitedError{}
 	}
-
-	for {
-		//Wait up to 1 second for Reads and Writes to the new connection
```

Comment:
Note to reviewer: this deadline actually means "when this time is reached, return with a temporary timeout error from accept", not what the comment says. There might have been a misunderstanding. This simply introduced an unnecessary loop every 1 second. It can be removed. The net/http Server already handles temporary timeouts and retries them, as specified in the PR description. Furthermore, they will now be logged too by the net/http package.

Reply:
I get what you're saying about the unnecessary retry at this level since a similar retry mechanism is handled at the net/http server level., but if you remove `SetDeadline` you fall back to the system level socket timeouts for connection establishment (I only see overrides in http server for read and write timeouts which deal with data after connection is already established). I seem to recall that this can lead to timeouts only triggering after as much as 60+ seconds.

Therefore, if we remove the call to `sl.SetDeadline` I think we'll be introducing the possibility of blocking the server on some weird connection state (this could be validated by adding a DROP rule via iptables for a particular source ip and check if the server still correctly accepts from other ips).

Reply:
My understanding was that the `SetDeadline` call means: if nobody connects until X, return an error from Accept. Why do we need this? As for the loop already exists here: https://github.com/golang/go/blob/go1.12/src/net/http/server.go#L2858-L2877

Or did you mean something else?

Reply:
So this isn't a timeout for a connection to be established in my understanding, it's a deadline after which Accept will return if nobody connects. Do you think it's not so?

Reply:
Ok, as discussed on Slack: this deadline could stop a potential TCP handshake timeout from stalling us forever, so I've added it back with an appropriate comment.